controllers/volumereplicationgroup_controller.go

```diff
@@ -1088,37 +1104,60 @@ func (v *VRGInstance) updateStatusState() {
 
 		v.log.Info("Failed to find the DataReady condition in status")
 
+		v.instance.Status.State = ramendrv1alpha1.UnknownState
+
 		return
 	}
 
 	StatusState := getStatusStateFromSpecState(v.instance.Spec.ReplicationState)
 
-	// update Status.State to reflect the state in spec
-	// only after successful transition of the resource
-	// (from primary->secondary or vise versa). That
-	// successful completion of transition can be seen
-	// in dataReadyCondition.Status being set to True.
-	if dataReadyCondition.Status == metav1.ConditionTrue {
+	if dataReadyCondition.Status != metav1.ConditionTrue ||
+		dataReadyCondition.ObservedGeneration != v.instance.Generation {
+		v.instance.Status.State = ramendrv1alpha1.UnknownState
+
+		return
+	}
+
+	if v.instance.Spec.ReplicationState == ramendrv1alpha1.Primary {
 		v.instance.Status.State = StatusState
 
 		return
 	}
 
-	// If VRG available condition is not true and the reason
-	// is Error, then mark Status.State as UnknownState instead
-	// of Primary or Secondary.
-	if dataReadyCondition.Reason == VRGConditionReasonError {
+	v.updateStatusStateForSecondary()
+}
+
+// updateStatusStateForSecondary is a helper to handle cases where VRG desired state is Secondary and its
+// status.State needs to be updated. See updateStatusState for more details.
+func (v *VRGInstance) updateStatusStateForSecondary() {
+	if v.instance.Spec.Action != ramendrv1alpha1.VRGActionRelocate {
+		v.instance.Status.State = ramendrv1alpha1.SecondaryState
+
+		return
+	}
+
+	dataProtectedCondition := findCondition(v.instance.Status.Conditions, VRGConditionTypeDataProtected)
+	if dataProtectedCondition == nil {
+		if len(v.volRepPVCs) == 0 {
+			// VRG is exclusively using volsync
+			v.instance.Status.State = ramendrv1alpha1.SecondaryState
+
+			return
+		}
+
 		v.instance.Status.State = ramendrv1alpha1.UnknownState
 
 		return
 	}
 
-	// If the state in spec is anything apart from
-	// primary or secondary, then explicitly set
-	// the Status.State to UnknownState.
-	if StatusState == ramendrv1alpha1.UnknownState {
-		v.instance.Status.State = StatusState
+	if dataProtectedCondition.Status == metav1.ConditionTrue &&
+		dataProtectedCondition.ObservedGeneration == v.instance.Generation {
```

Comment:
Should we ignore checking the `ObservedGenerator` here? For VolRep, the `DataProtected` condition will always be `false` until the time we are moving from primary to secondary. Said differently, once the `status` is set, it will not change and it **shouldn't** no matter what happens to the `generation`.  

The issue I am seeing is that there might be a possibility where we set the `state` to `secondary`, then the generation changes for any reason (before we clear the status), we find `VRGConditionTypeDataProtected` still exist, but the volRepPVCs count is 0, at that point, we reset the state to `Unknown`. Once that is done, the ObservedGeneration will never be updated in `aggregateVolRepDataProtectedCondition` due to `volRepPVCs` being 0.

It is not really a common use case, but it is probably we can avoid it by not testing for the `ObservedGeneration` here once the `status` condition is `true`.

Reply:
Correct, this condition not being nil causes some confusion in the Protected PR as well. Also the aggregation is not considering Unknown correctly afaict.

I am thinking of the following, for volsync report DataReady and DataProtected as secondary to be true always. Current report is nil so it does not get to vote in the condition, further the condition starts as Unknown and for pure volsync workloads it may remain as Unknown due to this (need to test).

See: https://github.com/RamenDR/ramen/pull/790/files#diff-c2546b038642973541a7ecf21f2d412877a0add33e7d37675635eb537ac262b7R129 and https://github.com/RamenDR/ramen/pull/790/files#diff-c2546b038642973541a7ecf21f2d412877a0add33e7d37675635eb537ac262b7R185

Thoughts?

Reply:
I think that would work.

Reply:
Added the change and retested VR and Volsync, both pass for a series or enable->failover->relocate->disable.

The aggregation though picks up the last true condition and that reports true with Unused (new reason in the code), need to correct it.

Otherwise this PR is good for a further review.